contrib/config/apollo/watcher.go

```diff
@@ -14,26 +18,40 @@ type customChangeListener struct {
 	event chan []*config.KeyValue
 }
 
-func (c *customChangeListener) OnChange(changeEvent *storage.ChangeEvent) {
-	kv := make([]*config.KeyValue, 0)
-	for key, value := range changeEvent.Changes {
-		kv = append(kv, &config.KeyValue{
-			Key:   key,
-			Value: []byte(value.NewValue.(string)),
-		})
+func (c *customChangeListener) onChange(
+	namespace string, changes map[string]*storage.ConfigChange) []*config.KeyValue {
+	kv := make([]*config.KeyValue, 0, 2)
+	next := make(map[string]interface{})
+
+	for key, change := range changes {
+		convertProperties(genKey(namespace, key), change.NewValue, next)
+	}
+
+	f := format(namespace)
+	codec := encoding.GetCodec(f)
+	val, err := codec.Marshal(next)
+	if err != nil {
+		_ = builtinLogger.Log(log.LevelWarn,
+			"msg",
+			fmt.Sprintf("apollo could not handle namespace %s: %v", namespace, err),
+		)
+		return nil
 	}
-	c.event <- kv
+	kv = append(kv, &config.KeyValue{
+		Key:    namespace,
+		Value:  val,
+		Format: f,
+	})
+
+	return kv
+}
+
+func (c *customChangeListener) OnChange(changeEvent *storage.ChangeEvent) {
+	c.event <- c.onChange(changeEvent.Namespace, changeEvent.Changes)
 }
 
 func (c *customChangeListener) OnNewestChange(changeEvent *storage.FullChangeEvent) {
-	kv := make([]*config.KeyValue, 0)
-	for key, value := range changeEvent.Changes {
-		kv = append(kv, &config.KeyValue{
-			Key:   key,
-			Value: []byte(value.(string)),
-		})
-	}
-	c.event <- kv
+	// TODO: finish this callback method. but it's not necessarily now.
```

Comment:
you'd better not include `TODO` in this PR if we don't need it, or could you please explain why don't need to implement it here?

Reply:
done